Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -12,311 +12,811 @@
+      highlightedSourceCode += decoratedHighlight
+
+      // If there's an additional line, append it.
+      if let decoratedAdditionalLine {
+        additionalHighlightedLine += decoratedAdditionalLine
+      } else {
+        additionalHighlightedLine += Array(repeating: " ", count: decoratedHighlight.count).joined()
       }
 
-      // line numbers should be right aligned
-      let lineNumberString = String(lineNumber)
-      let leadingSpaces = String(repeating: " ", count: maxNumberOfDigits - lineNumberString.count)
-      let linePrefix = "\(leadingSpaces)\(diagnosticDecorator.decorateBufferOutline("\(lineNumberString) │")) "
+      lastIndex = highlightIndexRange.upperBound
+    }
 
-      // If necessary, print a line that indicates that there was lines skipped in the source code
-      if hasLineBeenSkipped && !annotatedSource.isEmpty {
-        let lineMissingInfoLine = indentString + String(repeating: " ", count: maxNumberOfDigits) + " \(diagnosticDecorator.decorateBufferOutline("┆"))"
-        annotatedSource.append("\(lineMissingInfoLine)\n")
+    // Append any remaining part of the source string.
+    let remaining = sourceString[lastIndex...]
+    highlightedSourceCode += remaining
+    additionalHighlightedLine += Array(repeating: " ", count: remaining.count).joined()
+
+    return (
+      highlightedSourceCode: highlightedSourceCode,
+      /// Using `allSatisfy` to check if the `additionalHighlightedLine` consists only of whitespace characters.
+      /// If it does, we return `nil` to indicate that no additional highlighting exists.
+      additionalHighlightedLine: additionalHighlightedLine.allSatisfy(\.isWhitespace) ? nil : additionalHighlightedLine
+    )
+  }
+
+  /// Computes the set of highlight ranges for a given annotated source line.
+  ///
+  /// This function takes an `annotatedSourceLine` which contains diagnostics and returns an array of `Range<Int>` objects.
+  /// These ranges are sorted and merged to provide a concise collection of highlight regions for the source code line.
+  ///
+  /// - Parameters:
+  ///   - annotatedSourceLine: The line annotated with diagnostics.
+  ///   - tree: The syntax tree representing the entire source file.
+  ///   - sourceLocationConverter: The `SourceLocationConverter` used for converting between absolute and line/column positions.
+  ///
+  /// - Returns: An array of sorted and merged `Range<Int>` objects representing the highlight regions on the line.
+  private func computeHighlightRanges(
+    forLine annotatedSourceLine: AnnotatedSourceLine,
+    fromTree tree: some SyntaxProtocol,
+    usingSourceLocationConverter sourceLocationConverter: SourceLocationConverter
+  ) -> [Range<Int>] {
+    annotatedSourceLine.diagnostics
+      .compactMap {
+        switch $0 {
+        case let .diagnostic(diagnostic):
+          return diagnostic.highlights
+        case .note:
+          return nil
+        }
       }
```

Comment:
Instead of `compactMap {}.joined()` I think you can just do

```suggestion
      .flatMap {
        switch $0 {
        case let .diagnostic(diagnostic):
          return diagnostic.highlights
        case .note:
          return []
        }
      }
```

Reply:
Oh yeah, it will do the job. Thx.